okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java

```diff
@@ -155,14 +156,22 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
 
     @Override
     public void enqueue(Callback callback) {
-        AsyncTracer tracer = new AsyncTracer("OkHttp: acquire-limiter");
+        try (CloseableTracer ignored = CloseableTracer.startSpan("OkHttp: enqueue")) {
```

Comment:
Won't this lead to a very fast span (just enqueing) which is the parent of much longer spans (the actual call)? Ideally, the duration of a span would contain all its children. I would suggest either removing this, or having a span that actually lasts for the whole duration of the request (which I believe is something that #1069 does and was one of my original reasons to start discussing that change with @robert3005)

Reply:
> Ideally, the duration of a span would contain all its children

Ideally yes, but the more I think about it, the less important I think this is. Ideally a span's end time should:
1. Be after the end time of it's last child
2. Be before the start time of it's next sibling
However with async tasks, we can only pick one of these. I would argue that sibling spans occuring at the same time on the same trace is even more confusing than a parent that does not outlive it's child. Especially since that is exactly what is happening when work is done asynchronously - the parent finished submitting work before the work is done.

> I would suggest either removing this

We need something here under which all other request spans can exist. Otherwise all these spans are parented by the caller's span which leads to the confusion I detail in the PR description.

Reply:
The reason I think it is important is because if a span has a very small duration, I assume I don't need to expand it to check its children. It seems like a fair assumption to make, otherwise I would always have to expand all spans, no matter how long they are. I actually feel somewhat strongly about trying to keep a span's duration expanding through all its children duration.

If you have reasons not to remove this span, could you make it cover the whole request, from limiter to the actual response coming back.

Reply:
I don't disagree with you. This is actually the behavior I wanted when I started to make this changes. But doing this with the current API is very error-prone since it requires:
1. Calling `startSpan` at the beginning of the request
2. Then _not closing the span_ in order to keep in on the stack (this is different than the behavior of `AsyncTracer`, which immediately discards this span)
3. Calling `fastCompleteSpan` after the request is done, assuming that the span on the top of the stack is the one we started in step (1).

There are a couple things we can do to mitigate this:
- Design a better API to to safely support these types of spans (I think that's what https://github.com/palantir/tracing-java/pull/124 is attempting to do)
- Design a better API that verifies the span on the top of the stack is an expected span when closing

Reply:
Doing this actually requires removing the dispatcher span, otherwise (3) will never be true since the dispatcher span will always be on the top of the stack. But then you need another mechanism to transfer the trace from the submitting thread to the dispatcher thread.

Reply:
> Design a better API to to safely support these types of spans (I think that's what palantir/tracing-java#124 is attempting to do)

That's correct

> Design a better API that verifies the span on the top of the stack is an expected span when closing

Tracking here: https://github.com/palantir/tracing-java/issues/34

Reply:
Alternatively we could change the behavior of `DeferredTracer` to not create a span when an operation is not provided. Then the span stack would be in the correct state to complete the span after the request because there would be no dispatcher span.